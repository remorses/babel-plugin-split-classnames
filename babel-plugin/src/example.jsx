;<>
    <div>
        <Ciao
            className={clsx(
                'm-3 my-5 p-4 opacity-40 ',
                'md:w-80 m-3 my-5 p-4 opacity-40',
                'm-3 my-5 p-4 opacity-40 md:h-20',
            )}
        ></Ciao>
    </div>
</>
