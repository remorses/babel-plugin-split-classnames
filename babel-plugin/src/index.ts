export const DEFAULT_JSC_OPTIONS = {
    parser: 'tsx',
}
import {
    API,
    Collection,
    FileInfo,
    JSCodeshift,
    Options,
    JSXAttribute,
} from 'jscodeshift'

const CLASSNAMES_IDENTIFIER_NAME = 'cx'

const getClassNames = (string) => {
    return string
        .trim()
        .split(/\s/)
        .filter((name) => name.length > 0)
}

const isFalsyNodeValue = (node) =>
    (node.type === 'StringLiteral' && !node.value) ||
    node.type === 'NullLiteral' ||
    (node.type === 'Identifier' && node.name === 'undefined')

const _createObjectExpression = (j) => (entries) => {
    return j.objectExpression(
        entries.map(([property, value]) =>
            j.objectProperty.from({
                key: property,
                value,
                computed: true,
            }),
        ),
    )
}

const _createLiteral = (j) => (className) => {
    return j.literal(className)
}

const _createCxCallExpression = (j) => (args, classNamesIdentifier) => {
    return j.callExpression(j.identifier(classNamesIdentifier), args)
}

const _createImportDeclaration = (j) => (identifierName, source) => {
    return j.importDeclaration(
        [j.importDefaultSpecifier(j.identifier(identifierName))],
        j.stringLiteral(source),
    )
}

const _getLastLibImport = (j) => (ast) => {
    let firstImport = null
    let lastLibImport = null
    const importDeclarations = ast.find(j.ImportDeclaration)

    importDeclarations.forEach((path, i) => {
        const importSource = path.node.source.value
        if (i === 0) {
            firstImport = path
        }
        if (importSource.charAt(0) !== '.') {
            lastLibImport = path
        }
    })
    return lastLibImport || firstImport
}

const _getClassNamesIdentifierName = (j) => (ast) => {
    const importDeclarations = ast.find(j.ImportDeclaration, {
        type: 'ImportDeclaration',
        source: {
            value: 'classnames',
        },
    })

    if (importDeclarations.length === 1) {
        const importDeclaration = importDeclarations.get()
        const defaultImport = j(importDeclaration)
            .find(j.ImportDefaultSpecifier)
            .get()

        return defaultImport.node.local.name
    }
    return null
}

function splitClassNames(className: string, maxClassesPerGroup: number = 5) {
    const classes = className.split(/\s+/).filter((name) => name.length > 0)
    if (classes.length <= maxClassesPerGroup) {
        return [className.trim()]
    }
    const classGroups: string[] = []
    for (let i = 0; i < classes.length; i += maxClassesPerGroup) {
        classGroups.push(classes.slice(i, i + maxClassesPerGroup).join(' '))
    }
    return classGroups
}

export function transformer(
    fileInfo,
    api,
    options: Options & {
        classAttrName?: string
        classnamesImport?: string
        logicalExp?: boolean
        conditionalExp?: boolean
        falsyConditionalExp?: boolean
    },
) {
    try {
        const filePath = fileInfo.path
        const j: JSCodeshift = api.jscodeshift
        const createLiteral = _createLiteral(j)
        const createCxCallExpression = _createCxCallExpression(j)
        const createImportDeclaration = _createImportDeclaration(j)
        const createObjectExpression = _createObjectExpression(j)
        const getClassNamesIdentifierName = _getClassNamesIdentifierName(j)

        const ast: Collection = j(fileInfo.source)

        const transformLogicalExp = options.logicalExp
        const transformConditionalExpression = options.conditionalExp
        const transformFalsyConditionalExp = options.falsyConditionalExp
        const classAttrName = [
            'className',
            // ...(options.classAttrName || '').split(','),
        ]

        const existingClassNamesImportIdentifer =
            getClassNamesIdentifierName(ast)
        const classNamesImportName =
            existingClassNamesImportIdentifer ||
            options.classnamesImport ||
            CLASSNAMES_IDENTIFIER_NAME

        const lastLibImport: any = _getLastLibImport(j)(ast)
        let shouldInsertCXImport = false

        classAttrName.forEach((classAttrName) => {
            // simple literals or literals inside expressions
            ast.find(
                j.JSXAttribute,
                (attr: JSXAttribute) =>
                    attr.name.name === classAttrName &&
                    attr?.value?.type === 'StringLiteral',
            ).forEach((path) => {
                const literal = j(path).find(j.StringLiteral).get()

                const cxArguments = splitClassNames(literal.value?.value).map(
                    (s) => j.stringLiteral(s),
                )
                // don't add the classnames if className attr is short enough
                if (cxArguments.length <= 1) {
                    return
                }
                shouldInsertCXImport = true
                j(literal).replaceWith(
                    j.jsxExpressionContainer(
                        j.callExpression(
                            j.identifier(classNamesImportName),
                            cxArguments,
                        ),
                    ),
                )
            })
            // string literal inside expressions
            ast.find(
                j.JSXAttribute,
                (attr: JSXAttribute) =>
                    attr.name.name === classAttrName &&
                    attr?.value?.type === 'JSXExpressionContainer' &&
                    attr?.value?.expression?.type === 'StringLiteral',
            ).forEach((path) => {
                shouldInsertCXImport = true
                const literal = j(path).find(j.StringLiteral).get()

                const cxArguments = splitClassNames(literal.value?.value).map(
                    (s) => j.stringLiteral(s),
                )
                j(literal).replaceWith(
                    j.jsxExpressionContainer(
                        j.callExpression(
                            j.identifier(classNamesImportName),
                            cxArguments,
                        ),
                    ),
                )
            })

            // template literal
            ast.find(j.JSXAttribute, {
                type: 'JSXAttribute',
                name: {
                    type: 'JSXIdentifier',
                    name: classAttrName,
                },
                value: {
                    type: 'JSXExpressionContainer',
                    expression: {
                        type: 'TemplateLiteral',
                    },
                },
            }).forEach((path) => {
                shouldInsertCXImport = true
                const templateLiteral = j(path).find(j.TemplateLiteral).get()
                const { quasis, expressions } = templateLiteral.node
                let cxArguments: any[] = []
                quasis.forEach((quasi, index) => {
                    console.log('raw', quasi.value.raw)
                    if (quasi.value.raw.trim()) {
                        const classNames = splitClassNames(quasi.value.raw)
                        cxArguments.push(...classNames.map(createLiteral))
                    }
                    if (expressions[index] !== undefined) {
                        cxArguments.push(expressions[index])
                    }
                })
                j(templateLiteral).replaceWith(
                    j.callExpression(
                        j.identifier(classNamesImportName),
                        cxArguments,
                    ),
                )
            })

            // classnames arguments too long
            ast.find(j.JSXAttribute, {
                type: 'JSXAttribute',
                name: {
                    type: 'JSXIdentifier',
                    name: classAttrName,
                },
                value: {
                    type: 'JSXExpressionContainer',
                    expression: {
                        type: 'CallExpression',
                        callee: { name: classNamesImportName },
                    },
                },
            }).forEach((path) => {
                shouldInsertCXImport = true
                const callExpression = j(path).find(j.CallExpression).get()
                const newArgs: any[] = []
                console.log({ callExpression })
                callExpression.value.arguments.forEach((arg) => {
                    if (arg.type === 'StringLiteral') {
                        const newCxArguments = splitClassNames(arg.value).map(
                            (s) => j.stringLiteral(s),
                        )
                        newArgs.push(...newCxArguments)
                    } else {
                        newArgs.push(arg)
                    }
                })

                j(callExpression).replaceWith(
                    j.jsxExpressionContainer(
                        j.callExpression(
                            j.identifier(classNamesImportName),
                            newArgs,
                        ),
                    ),
                )
            })
        })

        if (!existingClassNamesImportIdentifer && shouldInsertCXImport) {
            lastLibImport.insertAfter(
                createImportDeclaration(classNamesImportName, 'classnames'),
            )
        }
        return ast.toSource(options as any)
    } catch (e) {
        console.error(e)
        throw e
    }
}
