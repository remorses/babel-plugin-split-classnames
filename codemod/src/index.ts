export const DEFAULT_JSC_OPTIONS = {
    parser: 'tsx',
}
import {
    API,
    Collection,
    FileInfo,
    JSCodeshift,
    Options,
    JSXAttribute,
} from 'jscodeshift'

const CLASSNAMES_IDENTIFIER_NAME = 'clsx'

function tailwindSort(a: string, b: string) {
    // a before b
    if (b.includes(':')) {
        return -1
    }
    // a after b
    if (a.includes(':')) {
        return 1
    }
    // a before b
    if (b.includes('[')) {
        return -1
    }
    // a after b
    if (a.includes('[')) {
        return 1
    }
    // a must be equal to b
    return 0
}

// TODO group classes by tailwind type (e.g. flex, grid, font and text, etc)
// groups are: defaults, md, lg, .etc, :dark, :hover, :focus, :active, :disabled
export function splitClassNames(
    className: string,
    maxClassLength: number = 60,
) {
    className = className.trim()
    if (className.length <= maxClassLength) {
        return [className]
    }
    const classes = className
        .split(/\s+/)
        .filter((name) => name.length > 0)
        .sort(tailwindSort)

    const classGroups: string[] = []
    let currentSize = 0
    let lastAddedIndex = 0

    for (let i = 0; i < classes.length; i += 1) {
        currentSize += classes[i].length
        if (currentSize >= maxClassLength || i === classes.length - 1) {
            classGroups.push(classes.slice(lastAddedIndex, i + 1).join(' '))
            lastAddedIndex = i + 1
            currentSize = 0
        }
    }

    return classGroups
}

// TODO support out of the box popular classnames libraries: clsx, classnames, etc.
// i can do this because i only change stuff inside className attribute, this is almost always a classnames implementation
// user can also give a priority implementation to use
// you can also use https://github.com/dcastil/tailwind-merge to merge tailwind stuff

// TODO let user choose if always add the clsx call instead of leaving short literals classes

const possibleClassNamesImportNames = new Set([
    'classnames',
    'classNames',
    'clsx',
    'cc',
    'cx',
    'cs',
    'classcat',
])

const CLASSNAMES_IMPORT_SOURCE = 'classnames'

export function transformer(
    fileInfo,
    api,
    options: Options & {
        classAttrNames?: string
        classnamesImport?: string
    },
) {
    try {
        const filePath = fileInfo.path
        const j: JSCodeshift = api.jscodeshift

        function createImportDeclaration(identifierName, source) {
            return j.importDeclaration(
                [j.importDefaultSpecifier(j.identifier(identifierName))],
                j.stringLiteral(source),
            )
        }

        const getClassNamesIdentifierName = (ast) => {
            const importDeclarations = ast.find(j.ImportDeclaration, {
                type: 'ImportDeclaration',
                source: {
                    value: CLASSNAMES_IMPORT_SOURCE,
                },
            })

            if (importDeclarations.length === 1) {
                const importDeclaration = importDeclarations.get()
                const defaultImport = j(importDeclaration)
                    .find(j.ImportDefaultSpecifier)
                    .get()

                return defaultImport.node.local.name
            }
            return null
        }

        const ast: Collection = j(fileInfo.source)

        const classAttrNames = [
            'className',
            ...(options.classAttrNames || '').split(','),
        ]
            .map((x) => x.trim())
            .filter(Boolean)

        const existingClassNamesImportIdentifier =
            getClassNamesIdentifierName(ast)
        const classNamesImportName =
            existingClassNamesImportIdentifier ||
            options.classnamesImport ||
            CLASSNAMES_IDENTIFIER_NAME

        const lastLibImport: any = (() => {
            let firstImport
            let lastLibImport
            const importDeclarations = ast.find(j.ImportDeclaration)

            importDeclarations.forEach((path, i) => {
                const importSource = path.node.source.value as string
                if (i === 0) {
                    firstImport = path
                }
                if (importSource?.charAt(0) !== '.') {
                    lastLibImport = path
                }
            })
            return lastLibImport || firstImport
        })()

        let shouldInsertCXImport = false

        for (const classAttrName of classAttrNames) {
            // simple literals or literals inside expressions
            ast.find(
                j.JSXAttribute,
                (attr: JSXAttribute) =>
                    attr.name.name === classAttrName &&
                    attr?.value?.type === 'StringLiteral',
            ).forEach((path) => {
                const literal = j(path).find(j.StringLiteral).get()

                const cxArguments = splitClassNames(literal.value?.value).map(
                    (s) => j.stringLiteral(s),
                )
                // don't add the classnames if className attr is short enough
                if (cxArguments.length <= 1) {
                    return
                }
                shouldInsertCXImport = true
                j(literal).replaceWith(
                    j.jsxExpressionContainer(
                        j.callExpression(
                            j.identifier(classNamesImportName),
                            cxArguments,
                        ),
                    ),
                )
            })
            // string literal inside expressions
            ast.find(
                j.JSXAttribute,
                (attr: JSXAttribute) =>
                    attr.name.name === classAttrName &&
                    attr?.value?.type === 'JSXExpressionContainer' &&
                    attr?.value?.expression?.type === 'StringLiteral',
            ).forEach((path) => {
                shouldInsertCXImport = true
                const literal = j(path).find(j.StringLiteral).get()

                const cxArguments = splitClassNames(literal.value?.value).map(
                    (s) => j.stringLiteral(s),
                )
                j(literal).replaceWith(
                    j.callExpression(
                        j.identifier(classNamesImportName),
                        cxArguments,
                    ),
                )
            })

            // template literal
            ast.find(j.JSXAttribute, {
                type: 'JSXAttribute',
                name: {
                    type: 'JSXIdentifier',
                    name: classAttrName,
                },
                value: {
                    type: 'JSXExpressionContainer',
                    expression: {
                        type: 'TemplateLiteral',
                    },
                },
            }).forEach((path) => {
                shouldInsertCXImport = true
                const templateLiteral = j(path).find(j.TemplateLiteral).get()
                const { quasis, expressions } = templateLiteral.node
                let cxArguments: any[] = []
                quasis.forEach((quasi, index) => {
                    if (quasi.value.raw.trim()) {
                        const classNames = splitClassNames(quasi.value.raw)
                        cxArguments.push(
                            ...classNames.map((className) =>
                                j.literal(className),
                            ),
                        )
                    }
                    if (expressions[index] !== undefined) {
                        cxArguments.push(expressions[index])
                    }
                })
                j(templateLiteral).replaceWith(
                    j.callExpression(
                        j.identifier(classNamesImportName),
                        cxArguments,
                    ),
                )
            })

            // classnames arguments too long
            ast.find(
                j.JSXAttribute,

                (attr: JSXAttribute) =>
                    attr.name.name === classAttrName &&
                    attr?.value?.type === 'JSXExpressionContainer' &&
                    attr?.value?.expression?.type === 'CallExpression' &&
                    possibleClassNamesImportNames.has(
                        // @ts-ignore
                        attr?.value?.expression?.callee?.name,
                    ),
            ).forEach((path) => {
                const callExpression = j(path).find(j.CallExpression).get()
                const newArgs: any[] = []
                const classNamesImportName = callExpression.value.callee.name
                callExpression.value.arguments.forEach((arg) => {
                    if (arg.type === 'StringLiteral') {
                        const newCxArguments = splitClassNames(arg.value).map(
                            (s) => j.stringLiteral(s),
                        )
                        newArgs.push(...newCxArguments)
                    } else {
                        newArgs.push(arg)
                    }
                })

                j(callExpression).replaceWith(
                    j.callExpression(
                        j.identifier(classNamesImportName),
                        newArgs,
                    ),
                )
            })
        }

        if (!existingClassNamesImportIdentifier && shouldInsertCXImport) {
            if (lastLibImport) {
                lastLibImport.insertAfter(
                    createImportDeclaration(
                        classNamesImportName,
                        CLASSNAMES_IMPORT_SOURCE,
                    ),
                )
            } else {
                ast.get().node.program.body.unshift(
                    createImportDeclaration(
                        classNamesImportName,
                        CLASSNAMES_IMPORT_SOURCE,
                    ),
                )
            }
        }
        return ast.toSource(options as any)
    } catch (e) {
        // console.error(e)
        throw e
    }
}
