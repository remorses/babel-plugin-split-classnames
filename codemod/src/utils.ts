import memoize from 'micro-memoize'
import prettier from 'prettier'

export function jscodeshiftToPrettierParser(parser) {
    const parserMap = {
        babylon: 'babel',
        flow: 'flow',
        ts: 'typescript',
        tsx: 'typescript',
    }

    return parserMap[parser] || 'babel'
}

export function warn(x) {
    return console.error('WARNING', x)
}

export const getPrettierConfig = memoize(() => {
    const prettierConfig = prettier.resolveConfig.sync('.', {
        editorconfig: true,
    }) || {
        printWidth: 100,
        tabWidth: 2,
        bracketSpacing: true,
        trailingComma: 'es5',
        singleQuote: true,
    }
    return prettierConfig
})

export function prettify(source) {
    const config = getPrettierConfig()
    if (!config) {
        console.warn(`Skipping code formatting as prettier config cannot be found`)
    }
    return prettier.format(source, {
        ...config,
        parser: 'typescript',
    })
}
