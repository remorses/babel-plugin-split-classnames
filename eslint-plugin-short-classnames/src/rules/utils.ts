import { ESLint } from 'eslint'

let eslint: ESLint
export async function runRule(code: string) {
    if (!eslint) {
        eslint = new ESLint({
            overrideConfig: {
                plugins: ['short-classnames'],
                parser: '@typescript-eslint/parser',
                rules: {
                    'short-classnames/short-classnames': [
                        'error',
                        { maxClassNameCharacters: 30 },
                    ],
                },
                parserOptions: {
                    ecmaVersion: 2018,

                    sourceType: 'module',
                    ecmaFeatures: {
                        jsx: true,
                    },
                },
            },
        })
    }
    const result = await eslint.lintText(code, { filePath: 'test.tsx' })

    if (!result[0]?.messages?.length) {
        return ''
    }
    let fixedCode = result[0].source
    let incrementRanges = 0
    for (let message of result[0].messages) {
        if (message.fix) {
            fixedCode = replaceRange(
                fixedCode,
                message.fix.range[0] + incrementRanges,
                message.fix.range[1] + incrementRanges,
                message.fix.text,
            )
            incrementRanges +=
                message.fix.text.length -
                (message.fix.range[1] - message.fix.range[0])
        }
    }
    return fixedCode
}

function replaceRange(s, start, end, substitute) {
    return s.substring(0, start) + substitute + s.substring(end)
}
