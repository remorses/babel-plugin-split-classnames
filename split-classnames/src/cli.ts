import fs from 'fs'
import path from 'path'
import { glob as globFn } from 'smart-glob'
import { runRule } from 'eslint-plugin-split-classnames/dist/utils'
import { Opts } from 'eslint-plugin-split-classnames/dist/rule'

import cac from 'cac'

const cli = cac(require('../package.json').name)

const allowedExtensions = ['.ts', '.tsx', '.js', '.jsx']

export async function runCodemod({ glob, opts = {} as Opts, dryRun = false }) {
    const files = await globFn(glob, {
        absolute: true,
        gitignore: true,

        ignoreGlobs: ['**/node_modules/**'],
    })

    const results: string[] = []
    for (let file of files) {
        let source = (await fs.promises.readFile(file)).toString()
        if (file.endsWith('.d.ts')) {
            continue
        }
        if (!allowedExtensions.some((ext) => file.endsWith(ext))) {
            continue
        }
        const ext = path.extname(file)
        console.info(`=> ${dryRun ? 'Found' : 'Applying to'} [${file}]`)
        source = (await runRule(source, opts)) || source
        results.push(source)
        if (!dryRun) {
            await fs.promises.writeFile(file, source, { encoding: 'utf-8' })
        }
    }
    return results
}

cli.command('[glob]', 'Split long classnames')
    .option('--dry', 'Only show what files would be changed', {
        type: ['boolean'],
    })
    .option('--max', 'Max number of characters in a classname')
    .action((glob, args) => {
        // console.log(args)

        if (!glob) {
            console.error('missing required positional argument glob')
            process.exit(1)
        }
        runCodemod({
            glob,
            opts: { maxClassNameCharacters: args.max || undefined },
            dryRun: args.dry,
        })
    })

cli.help()
cli.parse()

// console.log(JSON.stringify(argv, null, 2))
