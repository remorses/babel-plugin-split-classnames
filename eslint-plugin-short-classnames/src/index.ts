import { Rule } from 'eslint'
import { rule } from './rules/rule'

const plugin = {
    rules: {
        'short-classnames': rule,
    },

    configs: {
        parserOptions: {
            ecmaFeatures: {
                jsx: true,
            },
        },
        config: {
            plugins: ['short-classnames'],
            rules: {
                'short-classnames/short-classnames': 'error',
            },
        },
    },
}
export default plugin
module.exports = plugin
