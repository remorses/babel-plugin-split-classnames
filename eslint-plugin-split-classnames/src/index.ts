import { Rule } from 'eslint'
import path from 'path'
import { rule } from './rules/rule'

const plugin = {
    rules: {
        'split-classnames': rule,
    },

    configs: {
        parserOptions: {
            ecmaFeatures: {
                jsx: true,
            },
        },
        config: {
            plugins: ['split-classnames'],
            rules: {
                'split-classnames/split-classnames': 'error',
            },
        },
    },
}
export default plugin
module.exports = plugin
