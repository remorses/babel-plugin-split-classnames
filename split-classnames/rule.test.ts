import { ESLint, RuleTester } from 'eslint'
import { runRule } from 'eslint-plugin-split-classnames/src/utils'
import { test, describe, expect } from 'vitest'

const tests = {
    simple: `
    function Component() {
        return (
            <Fragment>
                <p
                    className='block w-[340px] sm:h-[60em] w-full py-2 mt-8 text-sm font-semibold text-center text-white bg-gray-900 border border-black rounded-md hover:bg-gray-700 disabled:bg-gray-900 hover:cursor-pointer disabled:opacity-60'
                />
            </Fragment>
        )
    }
    `,
    withTypescript: `
    function Component(x: SomeType) {
        return (
            <Fragment>
                <p
                    className='block w-[340px] sm:h-[60em] w-full py-2 mt-8 text-sm font-semibold text-center text-white bg-gray-900 border border-black rounded-md hover:bg-gray-700 disabled:bg-gray-900 hover:cursor-pointer disabled:opacity-60'
                />
            </Fragment>
        )
    }
    `,
    withManyClassnames: `
    function Component() {
        return (
            <Fragment>
                <p className='block w-[340px] sm:h-[60em] w-full py-2 mt-8 text-sm font-semibold text-center text-white bg-gray-900 border border-black rounded-md hover:bg-gray-700 disabled:bg-gray-900 hover:cursor-pointer disabled:opacity-60' />
                <p className='block w-[340px] sm:h-[60em] w-full py-2 mt-8 text-sm font-semibold text-center text-white bg-gray-900 border border-black rounded-md hover:bg-gray-700 disabled:bg-gray-900 hover:cursor-pointer disabled:opacity-60' />
            </Fragment>
        )
    }
    
    `,
    withImports: `
    import { Fragment } from 'react'
    function Component() {
        return (
            <Fragment>
                <p
                    className='block w-[340px] sm:h-[60em] w-full py-2 mt-8 text-sm font-semibold text-center text-white bg-gray-900 border border-black rounded-md hover:bg-gray-700 disabled:bg-gray-900 hover:cursor-pointer disabled:opacity-60'
                />
            </Fragment>
        )
    }
    `,
    withTemplateLiteral: `
    import { Fragment } from 'react'
    function Component() {
        return (
            <Fragment>
                <p
                    className={\`color classe foo \${true ? 'hello' : 'hi'} something py-2 mt-8 text-sm font-semibold text-center text-white\`}
                />
            </Fragment>
        )
    }
    `,
    callArgumentTooLong: `
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={cs('color classe foo something py-2 mt-8 text-sm font-semibold text-center text-white', true ? 'hello' : 'hi', 'something else')}
                />
            </Fragment>
        )
    }
    `,
    withCsImport: `
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={'color classe foo something py-2 mt-8 text-sm font-semibold text-center text-white'}
                />
            </Fragment>
        )
    }
    `,
    alreadySplitted: `
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={cs(
                        'color classe foo something py-2 mt-8 text-sm font-semibold',
                        'text-center text-white',
                        true ? 'hello' : 'hi',
                        'something else',
                        'wow so many',
                    )}
                />
            </Fragment>
        )
    }
    `,
    shouldNotRun: `
    import cs from 'classnames';
    function Component() {
    return (
            <Fragment>
            <p
                className={cs(
                'color classe foo something py-2 mt-8 text-sm font-semibold',
                'text-center text-white something else wow so many',
                true ? 'hello' : 'hi'
                )}
            />
            </Fragment>
        );
    }
    `,
    regression1: `
    import cs from 'classnames';
    function Component() {
    return (
            <Fragment>
            <p
                className={cs(
                    'appearance-none max-w-max flex whitespace-pre-wrap text-gray-700',
                    'pt-[1em] hover:underline dark:text-gray-100 first:pt-0',
                    hClasses[level] || '',
                    className,
                )}
            />
            </Fragment>
        );
    }
    `,
}

import prettier from 'prettier'

describe('test eslint', () => {
    for (let testName in tests) {
        test(`${testName}`, async () => {
            const code = tests[testName]
            // console.log('code', code)
            let fixedCode = await runRule(code)
            let prettyFixedCode =
                fixedCode &&
                prettier.format(fixedCode, {
                    singleQuote: true,
                    parser: 'babel',
                })
            // console.log(fixedCode)

            expect(prettyFixedCode).toMatchSnapshot('fixed')
            let fixedCodeAgain = (await runRule(fixedCode)) || fixedCode || code
            expect(fixedCodeAgain.trim()).toBe(fixedCode.trim())
        })
    }
})
