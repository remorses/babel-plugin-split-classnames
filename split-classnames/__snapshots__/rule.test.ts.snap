// Vitest Snapshot v1

exports[`test eslint > alreadySplitted > fixed 1`] = `
"
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={cs('color classe foo something py-2 mt-8 text-sm font-semibold', 'text-center text-white something else wow so many', true ? 'hello' : 'hi')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > callArgumentTooLong > fixed 1`] = `
"
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={cs('color classe foo something py-2 mt-8 text-sm font-semibold', 'text-center text-white something else', true ? 'hello' : 'hi')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > simple > fixed 1`] = `
"import clsx from 'classnames'

    function Component() {
        return (
            <Fragment>
                <p
                    className={clsx('block w-full py-2 mt-8 text-sm font-semibold text-center', 'text-white bg-gray-900 border border-black rounded-md', 'w-[340px] disabled:opacity-60 hover:cursor-pointer', 'disabled:bg-gray-900 hover:bg-gray-700 sm:h-[60em]')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > withCsImport > fixed 1`] = `
"
    import { Fragment } from 'react'
    import cs from 'classnames'
    function Component() {
        return (
            <Fragment>
                <p
                    className={cs('color classe foo something py-2 mt-8 text-sm font-semibold', 'text-center text-white')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > withImports > fixed 1`] = `
"import clsx from 'classnames'

    import { Fragment } from 'react'
    function Component() {
        return (
            <Fragment>
                <p
                    className={clsx('block w-full py-2 mt-8 text-sm font-semibold text-center', 'text-white bg-gray-900 border border-black rounded-md', 'w-[340px] disabled:opacity-60 hover:cursor-pointer', 'disabled:bg-gray-900 hover:bg-gray-700 sm:h-[60em]')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > withManyClassnames > fixed 1`] = `
"import clsx from 'classnames'

    function Component() {
        return (
            <Fragment>
                <p className={clsx('block w-full py-2 mt-8 text-sm font-semibold text-center', 'text-white bg-gray-900 border border-black rounded-md', 'w-[340px] disabled:opacity-60 hover:cursor-pointer', 'disabled:bg-gray-900 hover:bg-gray-700 sm:h-[60em]')} />
                <p className={clsx('block w-full py-2 mt-8 text-sm font-semibold text-center', 'text-white bg-gray-900 border border-black rounded-md', 'w-[340px] disabled:opacity-60 hover:cursor-pointer', 'disabled:bg-gray-900 hover:bg-gray-700 sm:h-[60em]')} />
            </Fragment>
        )
    }
    
    "
`;

exports[`test eslint > withTemplateLiteral > fixed 1`] = `
"import clsx from 'classnames'

    import { Fragment } from 'react'
    function Component() {
        return (
            <Fragment>
                <p
                    className={clsx('color classe foo ', true ? 'hello' : 'hi', 'something py-2 mt-8 text-sm font-semibold text-center', 'text-white')}
                />
            </Fragment>
        )
    }
    "
`;

exports[`test eslint > withTypescript > fixed 1`] = `
"import clsx from 'classnames'

    function Component(x: SomeType) {
        return (
            <Fragment>
                <p
                    className={clsx('block w-full py-2 mt-8 text-sm font-semibold text-center', 'text-white bg-gray-900 border border-black rounded-md', 'w-[340px] disabled:opacity-60 hover:cursor-pointer', 'disabled:bg-gray-900 hover:bg-gray-700 sm:h-[60em]')}
                />
            </Fragment>
        )
    }
    "
`;
